### PR TITLE
Feature Multiple Filtering

### PR DESCRIPTION
**Pull request by akshire**
# Multiple Filtering Enhancement
## TL;DR
Enhancements to Firebase Firestore enabling advanced multi-filtering capabilities and code refactoring.

### Improved User Experience
Based on user feedback, the `Apply` and `Reset` buttons have been swapped. Users reported that the new order feels more intuitive, improving usability.

Updated button order screenshot

*The Figma was consquently updated.*

### `ValueControl.kt` file
A new utils file was created to contain the `ValueRangeSlider` and the `CheckboxGroup` since these are reusable components for further UI implementations.

```kotlin
fun <T> CheckboxGroup(
    title: String,
    items: List<T>,
    selectedItem: StateFlow<T?>,
    onItemSelect: (T) -> Unit,
    testTagPrefix: String,
    emptyValue: T?
)

fun ValueRangeSlider(
    modifier: Modifier,
    name: String,
    min: Float,
    max: Float,
    range: StateFlow<FloatRange>,
    updateRange: (Float, Float) -> Unit,
)
```

### `TimeDisplay.kt` file
This utils file is supposed to contain every UI composable that should help in displaying time components.
A method was added to this file which is the `reformatTime` method
```kotlin
fun reformatTime(time: Float): String
```

### Firebase Firestore Enhancements
New Firestore indexes have been created to facilitate advanced filtering for `Recipe` objects. While unit testing for this feature is not feasible due to its dependence on the database, functionality can be verified within the app environment.

#### Documentation Update
A comprehensive wiki has been added, detailing how to integrate additional filtering features for `Recipe` objects, ensuring smoother onboarding for contributors.

https://github.com/PlateSwipe/PlateSwipe/wiki/Add-a-new-Filter-to-the-Swipe-Page-Firestore-version

Also documentation in the `filterSearch` function was updated with a link to the wiki.
## Why was this change made?
This update enables users to apply multiple filters simultaneously, enhancing the functionality and flexibility of the filtering system.

## Next Steps
To ensure robust testing, the full database needs to be regenerated with randomized values for time and difficulty for each `Recipe`. This will help validate edge cases. This task will be undertaken in collaboration with @RagdollGoDown and @Ylanv.

## Checklist

- [ ] Add unit tests (if applicable).
- [x] Update documentation:
    - Details about `filterSearch` added.
- [x] Pass all CI checks.
- [x] Link related issues or features: #268.

### CODE REVIEW

File: app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt [hunk 1]
```diff
@@ -305,17 +305,20 @@ class FirestoreRecipesRepository(private val db: FirebaseFirestore) : RecipeNetw
     var finalQuery: Query = db.collection(FIRESTORE_COLLECTION_NAME)
 
     /** ******************** Apply all the filters ********************* */
-
     /*
-     * Category filter
+    If you decide to change or add a filter in the future, you can add it here.
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt:309]: - **Documentation**: The added links to the wiki are great for onboarding! Just make sure the wiki stays updated. 👍

- **Category Filter**: Nice use of `.let`—clean and easy to follow. No issues here.

- **Time Filter**: `takeIf` is a solid choice for validation. Just double-check edge cases where `timeRange.min` or `max` hit the boundaries.

- **Difficulty Filter**: The `Difficulty.Undefined` check is clear and works well. Consistency across filters looks good too.

File: app/src/main/java/com/android/sample/model/filter/Filter.kt [hunk 2]
```diff
@@ -9,7 +9,13 @@ import com.android.sample.resources.C.Tag.Filter.UNINITIALIZED_BORN_VALUE
 import com.android.sample.resources.C.Tag.FilterPage.TIME_RANGE_MAX
 import com.android.sample.resources.C.Tag.FilterPage.TIME_RANGE_MIN
 
-/** Filter class to store filter types. */
+/**
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/filter/Filter.kt:12]: The new docs are super clear and well-written! Adding `@param` for each field is a great touch—makes it so much easier to follow. 🙌

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt [hunk 3]
```diff
@@ -3,63 +3,46 @@ package com.android.sample.ui.filter
...
 import com.android.sample.resources.C.Tag.FilterPage.TIME_RANGE_MAX
 import com.android.sample.resources.C.Tag.FilterPage.TIME_RANGE_MIN
-import com.android.sample.resources.C.Tag.MINUTES_PER_HOUR
 import com.android.sample.resources.C.TestTag.FilterPage.TEST_TAG_CATEGORY
 import com.android.sample.resources.C.TestTag.FilterPage.TEST_TAG_DIFFICULTY
 import com.android.sample.resources.C.TestTag.FilterPage.TEST_TAG_TIME_RANGE_SLIDER
 import com.android.sample.resources.C.TestTag.SwipePage.VIEW_RECIPE_BUTTON
 import com.android.sample.ui.navigation.NavigationActions
 import com.android.sample.ui.navigation.Screen
+import com.android.sample.ui.utils.CheckboxGroup
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:42]: - **Button Swap**: Swapping the `Reset` and `Apply` buttons based on user feedback is a solid move! It’s great to see user input being incorporated into the design. 👌

- **Code Extraction**: Moving utility functions like `ValueRangeSlider` and `CheckboxGroup`  to `ValueControl.kt` is definitely the right call. It makes `FilterPage.kt` lighter and improves code organization. Great for maintainability and reusability! 💡

File: app/src/main/java/com/android/sample/ui/utils/ValueControl.kt [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+import androidx.compose.ui.unit.dp
+import com.android.sample.model.filter.FloatRange
+import com.android.sample.resources.C.Dimension.PADDING_16
+import com.android.sample.resources.C.Dimension.PADDING_4
+import com.android.sample.resources.C.Dimension.PADDING_8
+import com.android.sample.resources.C.Tag.FilterPage.MAX_ITEM_IN_ROW
+import com.android.sample.resources.C.Tag.FilterPage.SLIDER_COLOR_ACTIVE
+import com.android.sample.resources.C.Tag.FilterPage.SLIDER_COLOR_INACTIVE
+import kotlinx.coroutines.flow.StateFlow
+
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/utils/ValueControl.kt:37]: - Splitting these utils into `ValueControl.kt` is a great idea—keeps the code modular and clean. 👍  
- `ValueRangeSlider` and `CheckboxGroup` are well-documented and reusable. No issues there. 👌  


Looks solid overall! 😊

File: app/src/main/java/com/android/sample/ui/utils/TimeDisplay.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import com.android.sample.R
+import com.android.sample.resources.C.Tag.MINUTES_PER_HOUR
+
+/**
+ * Composable function to reformat the time.
+ *
+ * @param time The time to reformat in minutes.
+ */
+@Composable
+fun reformatTime(time: Float): String {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/utils/TimeDisplay.kt:14]: - **Code Organization**: Moving `reformatTime` into a dedicated file is a good decision. It keeps utility functions centralized and reusable. 👍

- **Clarity**: The function is clear and well-documented. The use of constants (`MINUTES_PER_HOUR`) and `stringResource` makes the code readable and adaptable for localization. 👌

 **Suggestion**: Since this is a utility, you might consider writing a unit test to validate the formatting for edge cases (e.g., 0 minutes, exactly 60 minutes, and very large values).

**Review by 2Fick — COMMENTED** [review]
Great work on this PR! 🎉 The code is well-organized, and splitting the utility components into separate files makes everything cleaner and more reusable. Thanks for addressing the user feedback on the button swap—it’s a thoughtful improvement. 🙌

One small suggestion: consider adding unit tests for edge cases in the utility functions (`reformatTime` to ensure robustness).

Lastly, I noticed a potential issue: when applying a time filter and returning to the filter selection page, the time filter seems to reset automatically. Was this behavior intentional, or could it be an oversight? Let me know! 😊

Thanks again for the great work! 👏

File: app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt (hunk 1 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt:309]: Thanks

File: app/src/main/java/com/android/sample/model/filter/Filter.kt (hunk 2 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/filter/Filter.kt:12]: Thanks

File: app/src/main/java/com/android/sample/ui/filter/FilterPage.kt (hunk 3 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/ui/filter/FilterPage.kt:42]: Thanks

File: app/src/main/java/com/android/sample/ui/utils/ValueControl.kt (hunk 4 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/ui/utils/ValueControl.kt:37]: Thanks

**Comment by akshire** [comment]
The time resetting has been fixed thanks for reporting that!

File: app/src/main/java/com/android/sample/ui/utils/TimeDisplay.kt (hunk 5 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/ui/utils/TimeDisplay.kt:14]: Thanks for the comment, will do it right away !

File: app/src/androidTest/java/com/android/sample/utils/TimeDisplayTest.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package com.android.sample.utils
+
+import androidx.compose.ui.test.junit4.createComposeRule
+import com.android.sample.ui.utils.reformatTime
+import org.junit.Rule
+import org.junit.Test
+
+class TimeDisplayTest {
```
Comment (2Fick) [inline app/src/androidTest/java/com/android/sample/utils/TimeDisplayTest.kt:8]: Thank you for adding those tests as asked !

**Review by 2Fick — APPROVED** [review]
Great job addressing the changes and fixing the time filter bug! The updates look solid, and the code is well-organized and easy to follow. Thanks for your hard work on this! Approved! 🎉👏
